Marlin/MarlinSerial.cpp

```diff
@@ -101,24 +104,34 @@ void MarlinSerial::end() {
 
 
 int MarlinSerial::peek(void) {
-  if (rx_buffer.head == rx_buffer.tail) {
-    return -1;
-  }
-  else {
-    return rx_buffer.buffer[rx_buffer.tail];
-  }
+  CRITICAL_SECTION_START;
+    uint8_t t = rx_buffer.tail;
+    uint8_t c = rx_buffer.buffer[t];
+    if (rx_buffer.head == t) {
+  CRITICAL_SECTION_END;
+      return -1;
+    }
+    else {
+  CRITICAL_SECTION_END;
+      return c;
+    }
```

Comment:
The compiler is not allowed to optimize version 1 because rx_buffer.tail is volatile